src/java.base/share/classes/java/lang/invoke/VarHandles.java

```diff
@@ -217,14 +306,9 @@ static VarHandle makeArrayElementHandle(Class<?> arrayClass) {
         int ashift = 31 - Integer.numberOfLeadingZeros(ascale);
 
         if (!componentType.isPrimitive()) {
-            VarHandle vh;
-            if (UNSAFE.isFlatArray(arrayClass)) {
-                int layout = UNSAFE.arrayLayout(arrayClass);
-                vh = new VarHandleFlatValues.Array(aoffset, ashift, arrayClass, layout);
-            } else {
-                vh = new VarHandleReferences.Array(aoffset, ashift, arrayClass);
-            }
-            return maybeAdapt(vh);
+            // do not check for atomicity now -- atomicity will be checked on the accessed array object
+            // a sharper var handle will be obtained (dynamically) from flatArrayElementHandleFor(Object)
+            return maybeAdapt(new VarHandleReferences.Array(aoffset, ashift, arrayClass));
```

Comment:
Note: this works because:
* the reference Array plain access uses array syntax (which works no matter the array passed in)
* non-plain access modes always delegate to sharper var handles based on the type of the accessed array

Reply:
Another note: at the language level there's no mechanism to obtain a class mirror for a null-restricted, or atomic array. This means we can't really create an array var handle that works only on one array shape (which would allow us to eliminate the dynamic dispatch completely, at least for some array var handles).

Reply:
Yep, now it is crucial for us to benchmark these var handles - I suspect they are very sensitive to profile pollutions.